examples/memcached.md

```diff
@@ -0,0 +1,369 @@
+# memcached
+
+As shown in [this blog
+post](https://memcached.org/blog/persistent-memory/), memcached can
+efficiently utilize PMEM as replacement for DRAM. This makes it
+possible to increase cache sizes and/or reduce costs. The memcached
+maintainers make container images available which support this
+feature. This example shows how to take advantage of PMEM with
+memcached running on Kubernetes.
+
+## Prerequisites
+
+The instructions below assume that:
+- `kubectl` is available and can access the cluster.
+- PMEM-CSI [was installed](/docs/install.md#installation-and-setup)
+  with the default `pmem-csi.intel.com` driver name and with the
+  [`pmem-csi-sc-ext4` storage
+  class](/deploy/common/pmem-storageclass-ext4.yaml).
+- `telnet` or similar tools like `socat` are available.
+
+It is not necessary to check out the PMEM-CSI source code.
+
+## PMEM as DRAM replacement
+
+Memcached can map PMEM into its address space and then store the bulk
+of the cache there. The advantage is higher capacity and (depending on
+pricing) less costs.
+
+The cache could be stored in a persistent volume and reused after a
+restart (see next section), but when that isn't the goal, a
+[deployment with one CSI ephemeral inline
+volume](/deploy/kustomize/memcached/ephemeral/memcached-ephemeral.yaml)
+per memcached pod is very simple:
+- A Kubernetes deployment manages the memcached instance(s).
+- In the pod template, one additional PMEM volume of a certain size is
+  requested.
+- An init container prepares that volume for use by memcached:
+  - It determines how large the actual file can be that memcached will use.
+  - It changes volume ownership so that memcached can run as non-root
+    process.
+- A wrapper script adds the `--memory-file` and `--memory-limit` parameters
+  when invoking memcached, which enables the use of PMEM.
+
+In this mode, the PMEM-CSI driver is referenced through its name
+(usually `pmem-csi.intel.com`). No storage classes are needed. That
+name and other parameters in the deployment can be modified with
+[`kustomize`](https://github.com/kubernetes-sigs/kustomize). Here's
+how one can change the namespace, volume size or add additional
+command line parameters:
+
+```sh
+mkdir -p my-memcached-deployment
+
+cat >my-memcached-deployment/kustomization.yaml <<EOF
+namespace: demo
+bases:
+ - github.com/intel/pmem-csi/deploy/kustomize/memcached/ephemeral
+patchesStrategicMerge:
+- update-deployment.yaml
+EOF
+
+cat >my-memcached-deployment/update-deployment.yaml <<EOF
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: pmem-memcached
+spec:
+  template:
+    spec:
+      containers:
+      - name: memcached
+        env:
+        - name: MEMCACHED_ARGS
+          value: --conn-limit=500
+      volumes:
+      - name: data-volume
+        csi:
+          volumeAttributes:
+            size: 500Mi
+EOF
+```
+
+These commands then create the namespace and the deployment:
+```sh
+kubectl create ns demo
```

Comment:
I think for all the commands prepending with $ still makes sense.    `$ kubectl create ns demo`

Reply:
The problem with that is that it then becomes harder to cut-and-paste the command. As it is now (without the $), I can double-click on the command block to select all of it and the directly paste. That even works for multiple commands in a single block.

Ideally, the markup on the docsite should make it clearer which blocks contain code and which contain output. @intelkevinputnam: any suggestions how we can improve that?

See https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/view/pmem-csi/job/pmem-csi/view/change-requests/job/PR-633/Doc_20Site/examples/memcached.html

In Kubernetes, this looks like this:
https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/#creating-redis-deployment-and-service

Reply:
We addressed this problem with block labels in the [Clear Linux docs](https://docs.01.org/clearlinux/latest/guides/clear/mixer.html). I can do something similar with this. For example, I could put a "sh" label on input, and output could be labeled "output".

Reply:
You would be surprised how many times this discussion has come up in the docs meeting we have.  With GitHub Markdown you are limited in what you can do.  Almost all GitHub repo instructions I come across including kata all have the $ in front of the command.  It does make it harder to copy,  but it makes it more readable.  I've come to accept the $ because most people are doing instructions for the first or maybe second time I think they would prefer to do one command at a time.  I have my own personal way to remove all $ from a GitHub page that I am viewing since I expect it.

@intelkevinputnam At https://stackoverflow.com/questions/20303826/highlight-bash-shell-code-in-markdown I saw that "console" or "shell" can be used with code blocks and it supposedly colors the prompt. I've not tried it yet but that might make it even easier to read and follow.

Reply:
Using "ShellSession" in a fenced code block in markdown produces an interesting result when including $, however it doesn't fix the copy/paste problem.

What the markdown looks like.
````` md
``` ShellSession
$ make clean
Removing everything under '_build'...
$ make html
Running Sphinx v1.6.7
```
`````

What the output looks like: 
``` ShellSession
$ make clean
Removing everything under '_build'...
$ make html
Running Sphinx v1.6.7
```

It automatically colors the output differently from the input, keying off the inclusion of the $. Not sure how Sphinx/Read the Docs would handle it, though. I'll check.

Reply:
@intelkevinputnam I did some more digging and both   https://github.com/cri-o/cri-o and https://github.com/kubernetes/kubernetes skip using $ but https://github.com/containerd/containerd did use a $ to indicate bash code.  There is no consistency here so really it is whatever we want to do.

Reply:
I'd prefer to implement a solution that works equally well with the GitHub parser and Sphinx, unfortunately they use different lexers and the aliases for Shell sessions aren't the same. Markdown uses "ShellSession" and Sphinx uses "shell-session". So far, my preferred solution is to separate shell input and console output into different blocks and label them. I can split the difference fairly easily by adding a "$" to each line of a certain kind of code block via CSS that won't be picked up with select/copy.

Reply:
I take it back. Both GitHub markdown and Sphinx support the use of the alias "console" for a shell session to generate similar output with respect to using the $ character.

Reply:
So what's the conclusion?

Reply:
I have switched to using `console` and `$` because it is more consistent with the rest of the PMEM-CSI documentation (however, we still need to add `console` there).

IMHO, the increased readability is worth the loss of easy cut-and-paste:
- https://github.com/intel/pmem-csi/blob/a8041e3f3f5e355a688c2940d03869d273aaac97/examples/memcached.md
- https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/view/pmem-csi/job/pmem-csi/view/change-requests/job/PR-633/Doc_20Site/examples/memcached.html

Reply:
I skimmed through both of these and it does look a lot nicer to read.

Reply:
Looks good. I'm looking at an extension my colleague in Clear Linux docs found: sphinx_copybutton. We can add this extension and configure it with this:

copybutton_prompt_text = "$ "

This way the copy button that is appended to the code block will only pick up the commands and strip the "$" character.

Reply:
FYI, I just tested it on the current version of the Memcached page. Works like a charm. Block looks like this:
![image](https://user-images.githubusercontent.com/13109946/82238800-b2577000-98ec-11ea-9b6f-f05dbad6d8f1.png)
Clicking the copy button (top right) produces this:

mkdir -p my-memcached-deployment

cat >my-memcached-deployment/kustomization.yaml <<EOF

cat >my-memcached-deployment/update-deployment.yaml <<EOF

Reply:
Unfortunately, `cat >my-memcached-deployment/kustomization.yaml <<EOF` is incomplete: the lines up to and including `EOF` also need to be copied.

Let's merge this PR as it is and continue investigating this copy button.

Reply:
Ha, ha. Of course. I did a little tinkering and came up with a way to make this work.  Just requires a few lines of code in conf.py and some javascript. If you want to keep all of the lines and still strip the "$", we can use the "ShellSession" lexer alias. This way nothing changes in pure markdown, and we can use "console" when we use it for alternating input/output.

Reply:
I merged the memcached PR.

@intelkevinputnam: can you follow up with a PR that adds the copy-and-paste support? Do the other examples also need better annotations like `console`?

Reply:
Copy/Paste support with the extension adds it to all code blocks, so nothing else required. I'll put together some more explanation in the PR.